Nginx, staticfiles, IGV Reference files

JasonLing95 opened the pull request:
Changes:
1. changed logging in find_dx_data.py to print so that it shows up in cron.log through echo
2. changed python manage.py runserver to using gunicorn
3. web container no longer use port, just expose 8000 to internal Docker container
4. Added nginx container
5. minor update to example.env
6. new variable - PRIMER_DOWNLOAD (but still need a better way to automatically map hostname)
7. new variable - CSRF_TRUSTED_ORIGINS (for form submit purpose in production)

pep8speaks commented:
Hello @JasonLing95! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`ga_core/settings.py`](https://github.com/eastgenomics/Genetics_Ark/blob/9bbcd54d31edfe4f698a0d9a827bef17df6a4573/ga_core/settings.py):

> [Line 281:28](https://github.com/eastgenomics/Genetics_Ark/blob/9bbcd54d31edfe4f698a0d9a827bef17df6a4573/ga_core/settings.py#L281): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

JasonLing95 commented on docker-compose.yml:
```diff
@@ -2,22 +2,35 @@ version: '3.3'
-      - 8000:8000
+      - /var/run/docker.sock:/var/run/docker.sock # to call host docker
+      - /usr/bin/docker:/usr/bin/docker # to call host docker
+      - /home/jason/github/Genetics_Ark/tmp:/home/ga/tmp # zip generated primer designer PDFs
```
cause I currently testing it on my computer but it will require changing to server directory e.g. /home/lingj-loc/tmp or something if we test it on dev server

JasonLing95 commented on ga_core/settings.py:
```diff
@@ -26,10 +26,13 @@
 
     PROD_HOST = os.environ['PROD_HOST']
     DEBUG_HOST = os.environ['DEBUG_HOST']
+    CSRF_TRUSTED_ORIGINS = os.environ['CSRF_TRUSTED_ORIGINS']
```
for form submission purpose. The value for this variable is http://localhost:1337 currently which is used during testing as the server is running on localhost. During prod, this might be whichever hostname:port / url the prod is running

JasonLing95 commented:
# Updates
- igv browser loadDefaultGenomes set to `False`, instead load from local
- added shader for loading spinner gif
- minor edit to show warning modal if project name contains 'dev'
- Dockerfile now pull image from `primer_designer` instead of `python:3.8`
- cron container has its own Dockerfile which use python-slim image
- three docker volumes - static (for static files), primer_volume (for generated primer PDFs) and igv_volume (for the bam sample json file)
- added new variables in setting.py - `CSRF_TRUSTED_ORIGIN` (required for form submission in production), `CORS_ALLOW_ALL_ORIGINS` (will switch to `CORS_ALLOWED_ORIGINS` in production) - currently True to test on dev server but doesn't seems to solve the problem
- changed logging to two main logs (default / debug & error)
- added handler for page 404, 500 and 502
- limit primer design submission to 3 lines max
- updated dependencies in requirements.txt

JasonLing95 commented:
# Remaining Issue
- Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://dl.ec1.dnanex.us/F/D/jx2q7yz0pxFb9qB39pg9pzZjbxbjBvP2xgPqJ8P8/X221965-GM2214000-1-CEN-F-EGG5_markdup.bam.bai. (Reason: CORS request did not succeed).

JasonLing95 commented:
# Update
- Broken up `nexus_search` function in DNAnexus_to_igv/view.py to smaller individual function for easier reading
- Added pagination function to search function in DNAnexus_to_igv/view.py because if the search term is in many samples, we return a long list of samples previously as a scrollable list (which can be ~ 1GB) if search term is broad e.g. CEN. Now, we paginate sample results by 10 samples per page.

jethror1 commented on genetics_ark/templates/500.html:
```diff
@@ -0,0 +1,173 @@
+{% load static %}
```
what are these for?

JasonLing95 replied:
to load the files in staticfiles, 500.html need it for the font (FiraMono-Medium.ttf)

jethror1 replied:
sorry I meant more generally the new 404 and 500 templates, what do they look like?

JasonLing95 replied:
![Screenshot from 2022-07-27 09-02-23](https://user-images.githubusercontent.com/60819172/181194638-c55a74da-c9cf-4410-abc3-c6fb299012bc.png)

JasonLing95 replied:
Same image, but different errors: 404 Error, 500 Error & 502 Error. The "HOME" points to `/`

jethror1 commented on primer_designer/templates/primer_designer/index.html:
```diff
@@ -18,7 +39,7 @@
       <h2> Primer Designer</h2>
       <div>
-          Enter regions to design primers for (one region per line)
+          Enter regions to design primers for (one region per line - maximum 3)
```
any reason for adding a limit?

JasonLing95 replied:
Set it to 3 because currently gunicorn is set for a timeout of 600 which is ~ the time to generate 4 primers. More primer new line > longer processing time > cause workers to timeout thus the limit.

JasonLing95 commented:
# Pagination
- Reduce the amount of data transferred from backend to frontend at once - currently display only 10 samples per page
![Screenshot from 2022-07-27 09-04-37](https://user-images.githubusercontent.com/60819172/181195154-8d142ea3-b33c-439f-a907-7c1eca52d8ce.png)
